### PR TITLE
Feat(Link-expiry): Added link expiry using redis

### PR DESCRIPTION
**Pull request by Nazi-pikachu**
## Description
This draft pull request introduces the link expiry feature, allowing links to automatically expire after a specified period. 
The implementation includes adding a layer of redis on the top of postgres and Setting up TTL in redis

## Related Issue
https://github.com/Samagra-Development/yaus/issues/22
https://github.com/Samagra-Development/yaus/issues/52

### CODE REVIEW

**Comment by Nazi-pikachu** [comment]
Hey @yuvrajsab 
This is a draft PR with a minimal implementation of link-expiry using redis.
I am thinking of making RMQ durable Rather than non-durable since we can't afford data loss if we want to sync both the db using it.

**Comment by yuvrajsab** [comment]
@Nazi-pikachu yes please make it durable.

File: apps/api/src/app/app.service.ts [hunk 1]
```diff
@@ -12,27 +13,33 @@ export class AppService {
...
-  async fetchAllKeys(): Promise<string[]> {
-    const client = await this.redisService.getClient(this.configService.get<string>('REDIS_NAME'));
-    const keys: string[] = await client.keys('*');
-    return keys
-  }
+    /**
+     * Updates the click count in the postgres db based on hashId or customhashid
+     * @param hashID 
+     */
+    async updateClicksInPostgresDB(hashID: string): Promise<void> {
```
Comment (yuvrajsab) [inline apps/api/src/app/app.service.ts:24]: it should just be `id` bc we don't know now if it is hashid or customhashid.

Comment (yuvrajsab) [reply apps/api/src/app/app.service.ts:24]: and type should be number | string

File: apps/api/src/app/app.service.ts [hunk 2]
```diff
@@ -12,27 +13,33 @@ export class AppService {
...
-  async updateClicksInDb(): Promise<void> {
+    let cnt = await this.prisma.link.update({
+           where: { id: link.id },
+           data: { clicks: link.clicks + 1 }
+       });
+    }
+
+    // TO DO: shift to DB utils, required in scheduler !! so take care of that before shifting it to any other files
+    async updateClicksInDb(): Promise<void> {
     const client = await this.redisService.getClient(this.configService.get<string>('REDIS_NAME'));
```
Comment (yuvrajsab) [inline apps/api/src/app/app.service.ts:41]: remove this from scheduler if we are not using it

File: apps/api/src/app/app.controller.ts [hunk 3]
```diff
@@ -110,18 +112,19 @@ export class AppController {
...
   @ApiOperation({ summary: 'Update Existing Links' })
   @ApiBody({ type: Link })
   @ApiResponse({ type: Link, status: 200})
   async update(@Param('id') id: string, @Body() link: Link ): Promise<LinkModel> {
     return this.appService.updateLink({
-      where: { customHashId: id },
+      where: {customHashId: id},
       data: { 
         userID: link.user || null,
         tags: link.tags || null,
```
Comment (yuvrajsab) [inline apps/api/src/app/app.controller.ts:130]: update method implementation is not correct I think. it should be calling app service update method.

**Review by yuvrajsab — CHANGES_REQUESTED** [review]

File: apps/api/src/app/app.service.ts (hunk 1 above)
Comment (Nazi-pikachu) [reply apps/api/src/app/app.service.ts:24]: @yuvrajsab 
Here I just need to change the name of parameter and it's type 
Right ?

File: apps/api/src/app/app.controller.ts (hunk 3 above)
Comment (Nazi-pikachu) [reply apps/api/src/app/app.controller.ts:130]: @yuvrajsab 
Do you mean prisma update method as I cannot see any other method for update,
Also I have to modify the update link method to handle the case for I'd being custom hash I'd.
Also few edge cases where the user only want to change the link but not hashid is handled in it.

File: apps/api/src/app/app.service.ts (hunk 1 above)
Comment (Nazi-pikachu) [reply apps/api/src/app/app.service.ts]: resolved in 30ba88d124a83df18c841cfeb29b93247d842f49

**Comment by Nazi-pikachu** [comment]
This is final PR
It includes the following changes
1. Adding a layer of redis on top of postgres
2. Updating the relavant services to support this
3.  Integrating with RMQ to increase the cick_count in DB async
4. Refactoring the backend code to be more modular
5. Adding docString to make code more readable
6. Fixes few bugs and edge cases previously missing

**Review by yuvrajsab — APPROVED** [review]
lgtm.

**Pull request merged**